rplugin/python3/magma/images.py

```diff
@@ -130,12 +131,166 @@ def add_image(self, path: str, identifier: str, x: int, y: int, width: int, heig
             self._to_make_visible.add(identifier)
 
 
-def get_canvas_given_provider(name: str) -> Canvas:
-    """
-    Return a canvas object given its provider's name.
-    """
+# see https://sw.kovidgoyal.net/kitty/graphics-protocol/
+class KittyImage:
+    def __init__(self, id: int, path: str, row: int, col: int, width: int, height: int, nvim: Nvim):
+        self.id = id
+        self.path = path
+        self.row = row
+        self.col = col
+        self.width = width
+        self.height = height
+        self.nvim = nvim
+
+    def serialize_gr_command(self, **cmd):
+        payload = cmd.pop('payload', None)
+        cmd = ','.join('{}={}'.format(k, v) for k, v in cmd.items())
+        ans = []
+        w = ans.append
+        w(b'\033_G'), w(cmd.encode('ascii'))
+        if payload:
+            w(b';')
+            w(payload)
+        w(b'\033\\')
+        return b''.join(ans)
```

Comment:
see above..

Reply:
Ah, I didn't realise that you adapted code from there (I thought you were just linking to the reference of the protocol).

Then keep it as it is -- just change that comment to something like "Adapted from ..." instead of just "see ...".

Reply:
ok.